src/lib/cmdline.c

```diff
@@ -118,6 +118,141 @@ static int cbm_parse_cmdline_file_internal(const char *path, FILE *out)
         return nbytes;
 }
 
+/**
+ * Attempt to parse a command line file and remove its content from the out buffer.
+ *
+ * @Returns negative code if parsing failed, otherwise the new buffer size.
+ */
+static int cbm_parse_cmdline_file_removal_internal(const char *path, char *out, size_t buflen)
+{
+        autofree(FILE) *f = NULL;
+        size_t sn = 0;
+        ssize_t r = 0;
+        char *buf = NULL;
+        bool ret = true;
+        size_t nbytes = buflen;
+
+        f = fopen(path, "r");
+        if (!f) {
+                if (errno != ENOENT) {
+                        LOG_ERROR("Unable to open %s: %s", path, strerror(errno));
+                }
+                return false;
+        }
+
+        while ((r = getline(&buf, &sn, f)) > 0) {
+                ssize_t cur = 0;
+                char *m = NULL;
+                char *s = NULL;
+                size_t slen = 0;
+
+                /* Strip newlines */
+                if (r >= 1 && buf[r - 1] == '\n') {
+                        buf[r - 1] = '\0';
+                        --r;
+                }
+                char *l = buf;
+
+                /* Skip empty lines */
+                if (r < 1) {
+                        goto next_line;
+                }
+
+                /* Skip the starting whitespace */
+                while (isspace(*l)) {
+                        ++l;
+                        if (!*l) {
+                                break;
+                        }
+                }
+
+                /* Skip a comment */
+                if (l[0] == '#') {
+                        goto next_line;
+                }
+
+                /* Reset length */
+                cur = (l - buf);
+                r -= cur;
+
+                /* Strip trailing whitespace */
+                l = rstrip(l, (size_t)r, &r);
+
+                /* May now be an empty line */
+                if (r < 1) {
+                        goto next_line;
+                }
+
+                /* First look for a string with trailing whitespace */
+                s = string_printf("%s ", l);
+                slen = (size_t)(r + 1);
+                m = memmem(out, nbytes, s, slen);
```

Comment:
why not tokenize (`strtok(3)`) the cmdline and require matching the entire token instead? if i read the below correctly, it seems like it would 1) not remove multiple matches; 2) allow mutiple opt removals (maybe intentional), e.g. (cmdline)"opt1=val1 opt2 opt3" + (remove)"opt1=val1 opt22\n" => "opt3". it'll be cleaner, imo, and will remove endline as a special case.

Reply:
Ah I feel dumb for not even thinking about tokenizing. That said I did intend for 1 & 2s behavior (really I guess I should have added tests to verify that behavior though). I guess my thought was if you wanted to have command by command removal you could do newlines and if you wanted to get all or nothing you'd use the same line as kind of a way to say these options must be removed together. Add doesn't really need the same behavior because it will all be added regardless. Do you think that's a bad design and it would be just better to have the simple behavior of find and remove (though I'm inclined to keep the only remove first match behavior as that case seems weird to even happen).

Reply:
I think it's OK to have a it as it is. The testcases actually cover all the weird cases. I was just checking it was the intended behavior (+ special case of EOL bugs me which is special because a space is added forcibly to signify the end of token). I agree removing multiple matches should be problem of whoever put multiple occurrences there in the first place.